### ISSUE TITLE
Directly support openQASM circuits

### ISSUE THREAD

**Issue by nathanshammah** [unitaryfoundation/mitiq#1503]
## Issue Description

Mitiq could support directly as one of its frontends quantum circuits in openQASM format. Mitiq already has some conversion functions for openqasm wrt to Qiskit circuits.

## Additional References

@nickdgardner mentioned this as a useful feature.

**Comment by nickdgardner** [comment]
I've started looking at this (specifically for openQASM 3) and would be happy to collaborate on the issue with others who have more experience with Mitiq and/or openQASM.

**Comment by andreamari** [comment]
Interesting @nickdgardner !  This probably requires some planning/decisions. We can discuss more during a Mitiq call.

**Comment by natestemen** [comment]
Can someone lay out the value prop for this? From my understanding very few users are writing openqasm by hand, and many frontends can import/export from openqasm easily.

**Comment by nickdgardner** [comment]
I don't see a clear one now that I've looked into it more. But maybe @andreamari or @nathanshammah had something in mind? My initial thought was that it would be useful if Mitiq aims to eventually support people who want to apply mitigation methods involving sub-gate-level control, who might be working directly in QASM, and if QASM submodules would be the best way for Mitiq to implement those methods internally. But I see now this would be a huge undertaking, and I don't even know if it's a plausible scenario.

**Comment by andreamari** [comment]
QASM(2) is a [very popular](https://docs.google.com/presentation/d/e/2PACX-1vRjda57eKp7sLZENSJ7x83To2r0ja7XmW5sYaEobmYlvDKw122gagKm1j_qYm820054VbpFL0zZGjug/pub?start=false&loop=false&delayms=3000&slide=id.g14d46be9347_0_101) intermediate representation for circuits . 

Supporting QASM would make Mitiq easy to use with any software library that is able to export QASM circuits and any backend that is able to run QASM circuits. 

**Short term value prop:**
It would simplify the code of the user.
E.g., if Mitiq supported QASM as a fronted, there would be no need to manually convert circuits to Cirq in the [BSQkit tutorial](https://mitiq.readthedocs.io/en/stable/examples/bqskit.html).
Mitiq would do it internally.  So no fundamental advantages, just more user-friendly. 

**Long term value prop:**
In the long term, Mitiq could act directly on some fixed and unique intermediate representation (e.g. QASM2, QASM3 or QIR). This would  imply rewriting almost all the code of Mitiq which is instead Cirq-based, but it has some clear potential advantages. 

In practice, in my opinion, adding now QASM as a Mitiq frontend is perhaps not a super urgent issue but I see some short-term utility and some long-term potential.

**Comment by purva-thakre** [comment]
Forgot to add this comment:

Cirq 1.6 appears to support OpenQASM 3.0. https://github.com/quantumlib/Cirq/blob/48aadd033de005773c6f6374492b1fe48d40dfd3/cirq-core/cirq/circuits/circuit.py#L1344

However, Mitiq is not yet compatible with this version of Cirq. 

https://github.com/unitaryfund/mitiq/blob/b2de86a046a84211b4fce52994fdad7820939405/requirements/requirements-cirq.txt#L4

**Comment by cosenal** [comment]
@purva-thakre the latest version of Cirq that is released is 1.4.1, which is already supported by Mitiq.

**Comment by purva-thakre** [comment]
> [@purva-thakre](https://github.com/purva-thakre) the latest version of Cirq that is released is 1.4.1, which is already supported by Mitiq.

@cosenal Yes, thanks for pointing this out. This new compatibility is scheduled for a future release i.e. Cirq 1.6. I came across some comments regarding this in the cirq repo last week. I can't find them right now. Will update later.

**Comment by jordandsullivan** [comment]
If you wanted to just use `[qbraid.transpiler.transpile](https://docs.qbraid.com/sdk/user-guide/transpiler)`, you could support OpenQASM out of the box, as [UCC does](https://github.com/unitaryfoundation/ucc/blob/de56fbc22838b75c2d67b06c6f76678a8d705c66/ucc/compile.py#L3). Plus you get a all the other translations/conversions for free :]

**Comment by natestemen** [comment]
Can you help me understand where qbraid would be used in supporting openqasm? The simplest solution to this issue, IMO, is to add `str` to the potential `QPROGRAM` types, then use cirq's QASM importer ([`cirq.contrib.qasm_import.circuit_from_qasm`](https://quantumai.google/cirq/build/interop#importing_from_openqasm)) to convert the QASM to cirq. Once it's in cirq, do QEM, then export back to QASM with [`cirq.qasm`](https://quantumai.google/cirq/build/interop#exporting_cirqcircuit_to_qasm_format).

If this process works (always some unknown unknowns), I don't necessarily see where qbraid could help, unless we are talking a much large refactor to have all of mitiqs conversions go through qbraid.

**Comment by jordandsullivan** [comment]
If you didn't want to add qbraid conversions to all quantum circuits, you could add str as an accepted QPROGRAM type, and if a given program is a string, try to run it through `qbraid.transpile(circuit, 'cirq')` to convert to cirq and back to qasm `qbraid.transpile('qasm')`. 

But if you did want to just use `qbraid` to do all your conversions, that might save you  maintenance headache in the long-run (though indeed as you say it would require a much larger refactor in the immediate term).

**Comment by jordandsullivan** [comment]
I don't have super passionate feelings about using qbraid vs. cirq in this one instance, but just for completeness, here's my argument:

Again I would say the advantage to using qbraid to convert is that they deal with any API changes related to any and all conversions, and I think it's more likely that cirq would change the API in a lower-level module like [cirq.contrib.qasm_import.circuit_from_qasm](https://quantumai.google/cirq/build/interop#importing_from_openqasm) than qbraid would be to change how you call `qbraid.transpile`
